src/plugins/MediaManager.js

```diff
@@ -0,0 +1,345 @@
+            log.error('MediaDevices API unsupported: camera and microphone won\'t be available.');
+            this.instance.mediaEventsEnabled = false;
+            return;
+        }
+        this.instance.mediaEventsEnabled = true;
+
+        this.videoStreaming = false;
+        this.audioStreaming = false;
+
+        this.videoWidth = videoWidth;
+        this.videoHeight = videoHeight;
+        this.videoWithMicrophone = videoWithMicrophone;
+    }
+
+    /**
+     * Tries to get the current microphone and camera permission status. If it can, binds listeners to their change
+     * @returns {boolean} True if it could find the permissions object
+     */
+    async setupPermissions() {
+        if (!navigator.permissions) {
+            return false;
+        }
+
+        if (this.instance.options.microphone) {
+            try {
+                const permissionObj = await navigator.permissions.query({name: 'microphone'});
+                log.debug(`microphone ${permissionObj.state}`);
+                permissionObj.addEventListener('change', this.onMicrophonePermissionChange.bind(this));
+            } catch (error) {
+                log.warn('Can\'t get microphone permission object', error);
+                return false;
+            }
+        }
+
+        try {
+            const permissionObj = await navigator.permissions.query({name: 'camera'});
+            log.debug(`camera ${permissionObj.state}`);
+            permissionObj.addEventListener('change', this.onCameraPermissionChange.bind(this));
+        } catch (error) {
+            log.warn('Can\'t get camera permission object', error);
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Toggle local video forwarding.
+     * Redirect the client webcam video stream to the instance.
+     */
+    toggleVideoStreaming() {
+        if (!this.videoStreaming) {
+            this.startVideoStreaming();
+        } else {
+            this.stopVideoStreaming();
+        }
+    }
+
+    /**
+     * Toggle local audio forwarding.
+     * Redirect the client microphone audio stream to the instance.
+     */
+    toggleAudioStreaming() {
+        if (!this.audioStreaming) {
+            this.startAudioStreaming();
+        } else {
+            this.stopAudioStreaming();
+        }
+    }
+
+    /**
+     * Initialize and start client webcam video stream.
+     */
+    async startVideoStreaming() {
+        if (!navigator.mediaDevices) {
+            return;
+        }
+
+        try {
+            const mediaStream = await navigator.mediaDevices.getUserMedia({
+                audio: this.videoWithMicrophone,
+                video: {
+                    width: this.videoWidth,
+                    height: this.videoHeight,
+                },
+            });
+            log.debug('Client video stream ready');
+            this.videoStreaming = true;
+            this.localVideoStream = mediaStream;
```

Comment:
Should we set audio related fields as well if `this.videoWithMicrophone = true` ?  Or adapt toggleAudioStreaming to protected against starting an audio stream while mic is already captured?

Reply:
hm is it really an issue if the user sends a video&audio stream + an audio stream ?